lib/api/model/submessage.dart

```diff
@@ -0,0 +1,269 @@
+///
+/// https://zulip.com/api/get-messages#response
+@JsonSerializable(fieldRename: FieldRename.snake)
+class Submessage {
+  const Submessage({
+    required this.msgType,
+    required this.content,
+    required this.messageId,
+    required this.senderId,
+    required this.id,
+  });
+
+  @JsonKey(unknownEnumValue: SubmessageType.unknown)
+  final SubmessageType msgType;
+  /// [SubmessageData] encoded in JSON.
+  @JsonKey(readValue: readContent, toJson: contentToJson)
+  final Object? content;
+  final int messageId;
+  final int senderId;
+  final int id;
+
+  static Object? readContent(Map<Object?, Object?> json, String key) {
+    try {
+      return (const JsonDecoder().convert(json[key] as String)) as Object?;
+    } catch (e) {
+      if (e is FormatException) {
+        return null;
+      }
+      rethrow;
+    }
+  }
+
+  static Object? contentToJson(Object? content) =>
+    const JsonEncoder().convert(content);
+
+  factory Submessage.fromJson(Map<String, Object?> json) =>
+    _$SubmessageFromJson(json);
+
+  Map<String, Object?> toJson() => _$SubmessageToJson(this);
+}
+
+/// As in [Submessage.msgType].
+enum SubmessageType {
+  widget,
+  unknown,
+}
+
+sealed class SubmessageData {}
+
+/// The data encoded in a submessage to make the message a widget.
+///
+/// Expected from the first [Submessage.content] in the "submessages" field on
+/// the message when there is an widget.
+sealed class WidgetData extends SubmessageData {
+  WidgetType get widgetType;
+
+  WidgetData();
+
+  factory WidgetData.fromJson(Object? json) {
+    if (json is! Map) return UnsupportedWidgetData(json: json);
+    final map = json as Map<String, Object?>;
+    switch (WidgetType.fromJson(map['widget_type'] ?? 'unknown')) {
+      case WidgetType.poll: return PollWidgetData.fromJson(map);
+      default: return UnsupportedWidgetData(json: map);
+    }
+  }
+
+  Object? toJson();
+}
+
+/// As in [WidgetData.widgetType].
+@JsonEnum(alwaysCreate: true)
+enum WidgetType {
+  poll,
+  unknown;
+
+  factory WidgetType.fromJson(Object value) {
+    return $enumDecode<WidgetType, String>(
+      _$WidgetTypeEnumMap,
+      value,
+      unknownValue: WidgetType.unknown
+    );
+  }
+}
+
+/// The data encoded in a submessage to make the message a poll widget.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollWidgetData extends WidgetData {
+  @override
+  @JsonKey(includeToJson: true)
+  WidgetType get widgetType => WidgetType.poll;
+
+  /// The initial question and options on the poll.
+  final PollWidgetExtraData extraData;
+
+  PollWidgetData({required this.extraData});
+
+  factory PollWidgetData.fromJson(Map<String, Object?> json) =>
+    _$PollWidgetDataFromJson(json);
+
+  @override
+  Map<String, Object?> toJson() => _$PollWidgetDataToJson(this);
+}
+
+/// As in [PollWidgetData.extraData].
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollWidgetExtraData {
+  final String question;
+  final List<String> options;
```

Comment:
Yeah, those a good finds. Referencing https://github.com/zulip/zulip-flutter/pull/823#discussion_r1699090126, we can replicate the zulip mobile behavior by leaving `poll` as null as soon as the deserialization fails, so the message will be rendered into `ZulipContent`, instead of `PollContent`.

Reply:
I see; so we would keep `question` and `options` required—I agree with https://github.com/zulip/zulip-flutter/pull/823#discussion_r1699090126 that we can reasonably expect them to always be present—but also have error-handling, with a fallback to the usual message rendering style, in case of nasty surprises there (or anywhere)?

Reply:
I think it's best to match the existing clients' behavior — that simplifies away any attempt to verify that unexpected shapes of data never happen. I've added a commit to do that:
afd3c9b7e api: Give defaults to question/options on PollWidgetExtraData